internal/cri/server/images/image_pull.go

```diff
@@ -798,3 +838,182 @@ func (c *CRIImageService) snapshotterFromPodSandboxConfig(ctx context.Context, i
+	return &criCredentials{
+		ref:         ref,
+		credentials: credentials,
+	}
+}
+
+// GetCredentials gets credential from criCredentials makes criCredentials a registry.CredentialHelper
+func (cc *criCredentials) GetCredentials(ctx context.Context, ref string, host string) (registry.Credentials, error) {
+	if ref != cc.ref || cc.credentials == nil {
+		return registry.Credentials{}, nil
+	}
+
+	username, secret, err := cc.credentials(host)
+	if err != nil {
+		return registry.Credentials{}, nil
+	}
+	return registry.Credentials{
+		Host:     host,
+		Username: username,
+		Secret:   secret,
+	}, nil
+}
+
+type transferProgressReporter struct {
+	ref               string
+	pc                chan transfer.Progress
+	cancel            context.CancelFunc
+	timeout           time.Duration
+	reqReporter       pullRequestReporter
+	statuses          map[string]*transfer.Progress
+	lastSeenBytesRead uint64
+	lastSeenTimestamp time.Time
+}
+
+func newTransferProgressReporter(ref string, cancel context.CancelFunc, timeout time.Duration) *transferProgressReporter {
+	return &transferProgressReporter{
+		ref:      ref,
+		cancel:   cancel,
+		timeout:  timeout,
+		pc:       make(chan transfer.Progress),
+		statuses: make(map[string]*transfer.Progress),
+	}
+}
+
+func (reporter *transferProgressReporter) handleProgress(p transfer.Progress) {
+	// We only need to handle Progress Nodes that represent
+	// valid requests to a remote registry, so Progress nodes
+	// without 'Name', 'Desc' or 'Total' can be ignored
+	if p.Name == "" || p.Desc == nil || p.Total == 0 {
+		return
+	}
+
+	switch p.Event {
+	case "waiting":
+		// 'Waiting' events can be either when the layer is waiting to be
+		// downloaded and no progress has been made. Or when we have made
+		// some progress but `waiting` for more content to be downloaded.
+		if p.Progress == 0 {
+			return
+		}
+		fallthrough // Handle non-zero waiting progress same as downloading
+	case "downloading":
+		var curProgress int64
+		if node, ok := reporter.statuses[p.Name]; !ok {
+			curProgress = p.Progress
+			reporter.reqReporter.incRequest()
+		} else {
+			curProgress = p.Progress - node.Progress
+		}
+		reporter.statuses[p.Name] = &p
+		if curProgress > 0 {
+			reporter.UpdateBytesRead(curProgress)
+		}
+
+		// Download may be complete, but waiting for content
+		// to be written. In this case, we no longer consider it
+		// as an active requests.
+		if p.Progress == p.Total {
+			reporter.reqReporter.decRequest()
+			delete(reporter.statuses, p.Name)
+		}
+
+	case "complete":
+		if node, exists := reporter.statuses[p.Name]; exists {
+			if curProgress := p.Progress - node.Progress; curProgress > 0 {
+				reporter.UpdateBytesRead(curProgress)
+			}
+			reporter.reqReporter.decRequest()
+			delete(reporter.statuses, p.Name)
+		}
+	default:
+		return
+	}
+}
+
+func (reporter *transferProgressReporter) UpdateBytesRead(bytes int64) {
```

Comment:
suggest to use `incBytesRead` here. update sounds like setting an absolute value

Reply:
Updated